segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,457 @@
+        }
+
+        private EventsReadAndTruncationPoints deserializeEvents(BufferView inputData, long startOffset, int maxLength) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long nextOffset = startOffset;
+            final long maxOffset = startOffset + maxLength;
+            BufferView.Reader input = inputData.getBufferViewReader();
+            try {
+                while (nextOffset < maxOffset && events.size() < getConfig().getMaxItemsAtOnce()) {
+                    ProcessorEventData event = ProcessorEventSerializer.deserializeEvent(input);
+                    events.add(event);
+                    // Update the offset to the beginning of the next event.
+                    nextOffset += HEADER_LENGTH + event.getLength();
+                }
+            } catch (BufferView.Reader.OutOfBoundsException ex) {
+                // Events are of arbitrary size, so it is quite possible we stopped reading in the middle of an event.
+                // Only if we have already read some data, we do not throw.
+            }
+
+            if (events.isEmpty()) {
+                throw new NoDataAvailableException();
+            }
+            return new EventsReadAndTruncationPoints(events);
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> applyProcessorHandler(EventsReadAndTruncationPoints readResult) {
+            return getHandler().apply(readResult.getProcessorEventsData()).thenApply(v -> readResult);
+        }
+
+        private CompletableFuture<Void> truncateInternalSegment(EventsReadAndTruncationPoints readResult, Timer iterationTime) {
+            long truncationOffset = this.segment.getInfo().getStartOffset() + readResult.getReadBytes();
+            return this.segment.truncate(truncationOffset, containerOperationTimeout)
+                    .thenAccept(v -> {
+                        // Decrement outstanding bytes and publish the last latency value after successful truncation.
+                        this.outstandingBytes.addAndGet(-readResult.getReadBytes());
+                        this.lastIterationLatency.set(iterationTime.getElapsedMillis());
+                    });
+        }
+    }
+
+    @Data
+    private static class EventsReadAndTruncationPoints {
+        private final List<ProcessorEventData> eventsRead;
+
+        public long getReadBytes() {
+            return eventsRead.stream().map(d -> d.getLength() + HEADER_LENGTH).reduce(0, Integer::sum);
+        }
+
+        public List<BufferView> getProcessorEventsData() {
+            return eventsRead.stream().map(ProcessorEventData::getData).collect(Collectors.toUnmodifiableList());
+        }
+    }
+
+    private static class NoDataAvailableException extends RuntimeException {
+    }
+
+    /**
+     * Helper class to serialize/deserialize {@link ProcessorEventData} objects.
+     */
+    static class ProcessorEventSerializer {
```

Comment:
We should use VersionedSerializer here in the final production code

Reply:
Done.